optimizer/typeinferer.go

```diff
@@ -48,14 +51,111 @@ func (v *typeInferrer) Leave(in ast.Node) (out ast.Node, ok bool) {
+	case *ast.CompareSubqueryExpr:
+		x.SetType(types.NewFieldType(mysql.TypeLonglong))
+	case *ast.ExistsSubqueryExpr:
+		x.SetType(types.NewFieldType(mysql.TypeLonglong))
+	case *ast.PatternInExpr:
+		x.SetType(types.NewFieldType(mysql.TypeLonglong))
+	case *ast.PatternLikeExpr:
+		x.SetType(types.NewFieldType(mysql.TypeLonglong))
+	case *ast.PatternRegexpExpr:
+		x.SetType(types.NewFieldType(mysql.TypeLonglong))
+	case *ast.IsNullExpr:
+		x.SetType(types.NewFieldType(mysql.TypeLonglong))
+	case *ast.IsTruthExpr:
+		x.SetType(types.NewFieldType(mysql.TypeLonglong))
+	case *ast.ParenthesesExpr:
+		x.SetType(x.Expr.GetType())
+		// TODO: handle all expression types.
+	}
+	return in, true
+}
+
+func (v *typeInferrer) selectStmt(x *ast.SelectStmt) {
+	rf := x.GetResultFields()
+	for _, val := range rf {
+		if val.Column.ID == 0 && val.Expr.GetType() != nil {
+			val.Column.FieldType = *(val.Expr.GetType())
+		}
+	}
+}
+
+func (v *typeInferrer) binaryOperation(x *ast.BinaryOperationExpr) {
+	switch x.Op {
+	case opcode.AndAnd, opcode.OrOr, opcode.LogicXor:
+		x.Type = types.NewFieldType(mysql.TypeLonglong)
+	case opcode.LT, opcode.LE, opcode.GE, opcode.GT, opcode.EQ, opcode.NE, opcode.NullEQ:
+		x.Type = types.NewFieldType(mysql.TypeLonglong)
+	case opcode.RightShift, opcode.LeftShift, opcode.And, opcode.Or, opcode.Xor:
+		x.Type = types.NewFieldType(mysql.TypeLonglong)
+		x.Type.Flag |= mysql.UnsignedFlag
+	case opcode.IntDiv:
+		x.Type = types.NewFieldType(mysql.TypeLonglong)
+	case opcode.Plus, opcode.Minus, opcode.Mul, opcode.Mod:
+		if x.L.GetType() != nil && x.R.GetType() != nil {
+			xTp := mergeArithType(x.L.GetType().Tp, x.R.GetType().Tp)
+			x.Type = types.NewFieldType(xTp)
+			leftUnsigned := x.L.GetType().Flag & mysql.UnsignedFlag
+			rightUnsigned := x.R.GetType().Flag & mysql.UnsignedFlag
+			// If both operands are unsigned, result is unsigned.
+			x.Type.Flag |= (leftUnsigned & rightUnsigned)
+		}
+	case opcode.Div:
+		if x.L.GetType() != nil && x.R.GetType() != nil {
+			xTp := mergeArithType(x.L.GetType().Tp, x.R.GetType().Tp)
+			if xTp == mysql.TypeLonglong {
+				xTp = mysql.TypeDecimal
+			}
+			x.Type = types.NewFieldType(xTp)
+		}
+	}
+}
+
+func mergeArithType(a, b byte) byte {
+	switch a {
+	case mysql.TypeString, mysql.TypeVarchar, mysql.TypeVarString, mysql.TypeDouble, mysql.TypeFloat:
+		return mysql.TypeDouble
+	}
+	switch b {
+	case mysql.TypeString, mysql.TypeVarchar, mysql.TypeVarString, mysql.TypeDouble, mysql.TypeFloat:
+		return mysql.TypeDouble
+	}
+	if a == mysql.TypeNewDecimal || b == mysql.TypeNewDecimal {
+		return mysql.TypeNewDecimal
+	}
+	return mysql.TypeLonglong
+}
+
+func (v *typeInferrer) unaryOperation(x *ast.UnaryOperationExpr) {
+	switch x.Op {
+	case opcode.Not:
+		x.Type = types.NewFieldType(mysql.TypeLonglong)
+	case opcode.BitNeg:
+		x.Type = types.NewFieldType(mysql.TypeLonglong)
+		x.Type.Flag |= mysql.UnsignedFlag
+	case opcode.Plus:
+		x.Type = x.V.GetType()
```

Comment:
Plus works differently than Minus, `select +'abc'` returns the original varchar type.